Content.Shared/Weapons/Melee/MeleeThrowOnHitSystem.cs

```diff
@@ -15,113 +16,72 @@ namespace Content.Shared.Weapons.Melee;
+            ThrowOnHitHelper(weapon, args.User, target, direction);
         }
     }
 
-    private void OnThrownStartup(Entity<MeleeThrownComponent> ent, ref ComponentStartup args)
+    private void OnThrowHit(Entity<MeleeThrowOnHitComponent> weapon, ref ThrowDoHitEvent args)
     {
-        var (_, comp) = ent;
+        if (!weapon.Comp.ActivateOnThrown)
+            return;
 
-        if (!TryComp<PhysicsComponent>(ent, out var body) ||
-            (body.BodyType & (BodyType.Dynamic | BodyType.KinematicController)) == 0x0)
+        if (!TryComp<PhysicsComponent>(args.Thrown, out var weaponPhysics))
             return;
 
-        comp.PreviousStatus = body.BodyStatus;
-        comp.ThrownEndTime = _timing.CurTime + TimeSpan.FromSeconds(comp.Lifetime);
-        comp.MinLifetimeTime = _timing.CurTime + TimeSpan.FromSeconds(comp.MinLifetime);
-        _physics.SetBodyStatus(ent, body, BodyStatus.InAir);
-        _physics.SetLinearVelocity(ent, Vector2.Zero, body: body);
-        _physics.ApplyLinearImpulse(ent, comp.Velocity * body.Mass, body: body);
-        Dirty(ent, ent.Comp);
+        ThrowOnHitHelper(weapon, args.Component.Thrower, args.Target, weaponPhysics.LinearVelocity);
     }
 
-    private void OnThrownShutdown(Entity<MeleeThrownComponent> ent, ref ComponentShutdown args)
+    private void ThrowOnHitHelper(Entity<MeleeThrowOnHitComponent> ent, EntityUid? user, EntityUid target, Vector2 direction)
     {
-        if (TryComp<PhysicsComponent>(ent, out var body))
-            _physics.SetBodyStatus(ent, body, ent.Comp.PreviousStatus);
-        var ev = new MeleeThrowOnHitEndEvent();
-        RaiseLocalEvent(ent, ref ev);
-    }
+        var attemptEvent = new AttemptMeleeThrowOnHitEvent(target, user);
+        RaiseLocalEvent(ent.Owner, ref attemptEvent);
 
-    private void OnStartCollide(Entity<MeleeThrownComponent> ent, ref StartCollideEvent args)
-    {
-        var (_, comp) = ent;
-        if (!args.OtherFixture.Hard || !args.OtherBody.CanCollide || !args.OurFixture.Hard || !args.OurBody.CanCollide)
+        if (attemptEvent.Cancelled)
             return;
 
-        if (_timing.CurTime < comp.MinLifetimeTime)
-            return;
+        var startEvent = new MeleeThrowOnHitStartEvent(ent.Owner, user);
+        RaiseLocalEvent(target, ref startEvent);
 
-        RemCompDeferred(ent, ent.Comp);
-    }
-
-    public bool CanThrowOnHit(Entity<MeleeThrowOnHitComponent> ent, EntityUid target)
-    {
-        var (uid, comp) = ent;
+        if (ent.Comp.StunTime != null)
+            _stun.TryParalyze(target, ent.Comp.StunTime.Value, false);
+        if (ent.Comp.UnanchorOnHit && HasComp<AnchorableComponent>(target))
+            _transform.Unanchor(target);
 
-        var ev = new AttemptMeleeThrowOnHitEvent(target);
-        RaiseLocalEvent(uid, ref ev);
-
-        if (ev.Handled)
-            return !ev.Cancelled;
-
-        return comp.Enabled;
-    }
+        if (direction == Vector2.Zero)
+            return;
```

Comment:
This should really be an extension of throwing given how generic is rather than having each system having to handle it.

Reply:
Yeah, the stun part is prolly fine.